test_apps/datamgr/test_datamgr.cpp

```diff
@@ -156,35 +157,46 @@ void test_get_value(Grid *g)
 
     g->SetInterpolationOrder(1);
 
-    Grid::ConstIterator itr = g->cbegin();
-    Grid::ConstIterator enditr = g->cend();
-
-    Grid::ConstCoordItr c_itr = g->ConstCoordBegin();
-    Grid::ConstCoordItr c_enditr = g->ConstCoordEnd();
-
     const float epsilon = 0.000001;
 
-    float  t0 = GetTime();
-    size_t ecount = 0;
-    for (; itr != enditr; ++itr, ++c_itr) {
-        float v0 = *itr;
+    float t0 = GetTime();
 
-        float v1 = g->GetValue(*c_itr);
+    size_t n = VProduct(g->GetDimensions());
 
-        if (v0 != v1) {
-            if (v0 == 0.0) {
-                if (abs(v1) > epsilon) {
-                    ecount++;
-                    v1 = g->GetValue(*c_itr);
-                }
-            } else {
-                if (abs((v1 - v0) / v0) > epsilon) {
-                    ecount++;
-                    v1 = g->GetValue(*c_itr);
+    size_t ecount = 0;
+#pragma omp parallel
```

Comment:
this parallel region doesn't seem to honor the global num. of thread setting. Recommend specifying it by using the num_threads clause:
`#pragma omp parallel num_threads( N )`

The same comment applies to other parallel regions.

Reply:
Fixed.

Reply:
did you push your change? I don't see any QuadTree parallel regions honor the num. of threads setting. 

Also, I'm not sure if it's a good idea to put a function call in a pragma clause. The standard doesn't seem to specify (https://www.openmp.org/spec-html/5.0/openmpse14.html#x54-800002.6). I would do the following: 

```c++
int nthreads = VAPOR::GlobalParams::Get_user_setting_nthreads();
#pragma omp parallel num_threads( nthreads ) {}
```

Reply:
good point. fixed.

Reply:
I don't quite understand what your argument is, but my bigger point is that 
1) If VAPOR GUI provides a setting of num. of threads, then the application should honor it, no matter what the environment variable is or even there's no such environment variable set. 
2) Alternatively we can simply remove the GUI setting of num. of threads, and solely depend on the environment variable to control the number of threads. 

I agree that it's not a big deal in a test program, but I do think it's important to have the real application honor its own settings.

Reply:
Understood. I'd forgotten that VAPOR GUI allows the specification of the number of threads as a parameter. I think the GUI parameter is probably more intuitive than expecting the  users to know about implementation details of OpenMP. I'd propose simply having the GUI set the OMP_NUM_THREADS env. variable based on the parameter setting. Thus both the number of PThreads and the number of OpenMP threads would be set by this setting. If there were ever a reason to allow PThreads and OpenMP threads separately we could simply add another parameter. Thoughts on this?

Reply:
I dont see why you need to involve in the `OMP_NUM_THREADS` EV: it's a pretty unreliable way to specify something: so many different players can go change the value of an EV, for example, the job scheduler. I propose just ask both pthread or OpenMP to grab the value that's set by the user and apply that value as the num. of threads.

In the case of user sets 0 as the num of thread, on the OpenMP part, we can ask OpenMP to use whatever it likes (which could be the value of `OMP_NUM_THREADS` if this EV is defined).  That's the logic of my suggestion:
```c++
int nthreads = VAPOR::GlobalParams::Get_user_setting_nthreads();
if( nthreads == 0 ) {
    nthreads == omp_get_num_threads();
}
#pragma omp parallel num_threads( nthreads ) {}
```

Reply:
An environment variable set by a unix process will overwrite any previous settings. If we do it this way we don't have to have special handling for every pragma omp parallel.

Reply:
If you put the logic into the OpenMP helper header, then it's just one line to retrieve the number of threads from the header, which I think is a small enough burden. Additionally, environment variables on Windows is probably less reliable than Unix.

Imagine a user sets `OMP_NUM_THREADS` to use SZ to compress some data. Then he briefly uses VAPOR, and then returns to his SZ compression work. Now the value of `OMP_NUM_THREADS` has changed without his knowledge! (This is bad.)

My intuition is that if something you can do within the scope of your application, then do it within the scope of your application and not mess up the outside world.

Reply:
What you suggest is probably fine, but your understanding of environment variables is not correct; an environment variable set within a process is not visible outside of that process. I.e. it doesn't change the value of the environment variables in the parent shell that spawned the process. Moreover, OpenMP already has a mechanism for setting the number of threads - enviorment variables - why do we need to create a new one?

Reply:
Oh Ok, if it doesn't interfere other applications, then I'm fine with it. 

OpenMP has not one, but 3 mechanisms for setting the num. of threads, so nobody is creating a new one. 

Among these 3, an Envrionment Variable has the lowest priority, and a clause appended to a `#pragma omp parallel` has the highest priority (when these two are in conflict, the value appeared in the clause will be used). I think when you know the num. of threads to use, there's merit to use the highest priority mechanism to specify it.

Reply:
As it turns out there was no reason to use an environment variable; just call omp_set_num_threads(). Fixed.

Reply:
Looks good!